### ISSUE TITLE
GeometryFixer throws `TopologyException` on some geometries

### ISSUE THREAD

**Issue by mjanuszkiewicz-tt** [locationtech/jts#852]
We have encountered some geometries that cause `GeometryFixer` to throw `TopologyException`s:

https://gist.github.com/mjanuszkiewicz-tt/7c5313310eb6a3a15bd000e0735d7606

The exceptions are
`org.locationtech.jts.geom.TopologyException: side location conflict: arg 1 [ (50.69759761418551, 4.590501980547396, NaN) ]`
and
`org.locationtech.jts.geom.TopologyException: unable to assign free hole to a shell [ (42.5631796905326, -72.61259223074235, NaN) ]`

The problem is that the internal `fixRing` method used to fix holes can produce geometry that is still invalid.
We patched the `fixedHoles` method locally to the effect of:
```
if(holeRep != null && !holeRep.isValid()) {
    holeRep = fix(holeRep);
}
```

This is good enough for us, but I suspect not in the general case.

**Comment by dr-jts** [comment]
The problem is that the `BufferOp.bufferByZero(poly, true)` technique used in `GeometryFixer.fixRing` is failing to produce a valid polygonal result, due to two intersecting line segments with a vertex lying extremely close to the other segment.

Here's a minimum reproducer.  This is a portion of a hole in case 1 above.  This is invalid due to a self-intersection. 
 When this is processed with `GeometryFixer`, it does not produce a valid result.  This causes the union of the holes in the original case to fail.
```
POLYGON ((42.565844354657436 -72.61247966084643, 42.56484510561062 -72.61202938126273, 42.56384585656381 -72.61247966084643, 42.563637679679054 -72.61276108558623, 42.562055535354936 -72.61366164475362, 42.5631796905326 -72.61259223074235, 42.565844354657436 -72.61214195115866, 42.566510520688645 -72.61259223074235, 42.565844354657436 -72.61247966084643))
```
![image](https://user-images.githubusercontent.com/3529053/167228257-27fbb44b-3d51-4e83-8727-b5a1366d31af.png)


The invalid result from `GeometryFixer`.  The two final points of the final shell are almost identical, and they create a self-intersection in the shell.  This is fundamentally a problem with the noding used in `buffer` (a common cause of robustness problems).
```
MULTIPOLYGON (((42.56529930972282 -72.61223405380078, 42.565844354657436 -72.61214195115866, 42.566510520688645 -72.61259223074235, 42.565844354657436 -72.61247966084643, 42.56529930972282 -72.61223405380078)), ((42.56384585656382 -72.61247966084643, 42.56484510561062 -72.61202938126273, 42.56529930972282 -72.61223405380078, 42.56384585656382 -72.61247966084643)), ((42.56384585656382 -72.61247966084643, 42.5631796905326 -72.61259223074235, 42.562055535354936 -72.61366164475362, 42.563637679679054 -72.61276108558623, 42.56384585656381 -72.61247966084643, 42.56384585656382 -72.61247966084643)))
```
![image](https://user-images.githubusercontent.com/3529053/167228272-e3575ada-00b0-4c68-b3f7-eade8ba6d0e1.png)

**Comment by dr-jts** [comment]
Here's a reproducer from Case 2.  It is a hole in the original geometry, which produces an invalid result from `fixRing`, and thus causes the subtraction of holes to fail.
```
POLYGON ((50.69544005538049 4.587126197745181, 50.699035986722194 4.592752502415541, 50.699395579856365 4.592049214331746, 50.699125885005735 4.590501980547397, 50.69867639358802 4.591064611014433, 50.69795720731968 4.591064611014433, 50.69759761418551 4.590501980547397, 50.69759761418551 4.589376719613325, 50.69831680045385 4.588251458679252, 50.69723802105134 4.586563567278144, 50.69579964851466 4.586563567278144, 50.69544005538049 4.587126197745181))
```

**Comment by dr-jts** [comment]
The fundamental problem is that the noding used in the buffer algorithm is non-robust, and thus can produce incorrect topology.  In the overlay code this is handled by checking the correctness of the noding after it is created.  This is effective, but does impact performance. This is not done in the buffer algorithm, probably because buffer topology building catches noding errors later on.  It hasn't caused any known problems - until this issue.  

A fix is to check the noding done in buffer.  This can be done for `distance=0` only, to avoid impacting performance of other buffers.

**Comment by dr-jts** [comment]
There is a fix in #867.  @mjanuszkiewicz-tt are you able to test this?
